File: androidApp/src/main/kotlin/com/weather/android/ui/molecules/WeatherDetailCard.kt
package com.weather.android.ui.molecules

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.weather.domain.model.Weather
import com.weather.domain.model.AirQualityCategory

@Composable
fun WeatherDetailCard(
    title: String,
    icon: ImageVector,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit
) {
    Card(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = 4.dp)
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.padding(bottom = 12.dp)
            ) {
                Icon(
                    imageVector = icon,
                    contentDescription = title,
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.size(24.dp)
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            content()
        }
    }
}

@Composable
fun TemperatureDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    WeatherDetailCard(
        title = "Temperature",
        icon = Icons.Default.Add,
        modifier = modifier
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            weather.temperatureCurrent?.let { current ->
                DetailRow("Current", "${current.toInt()}°C", isHighlight = true)
            }
            weather.feelsLike?.let { feelsLike ->
                DetailRow("Feels like", "${feelsLike.toInt()}°C")
            }
            DetailRow("High", "${weather.temperatureHigh.toInt()}°C")
            DetailRow("Low", "${weather.temperatureLow.toInt()}°C")
            weather.dewPoint?.let { dewPoint ->
                DetailRow("Dew point", "${dewPoint.toInt()}°C")
            }
        }
    }
}

@Composable
fun WindDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    WeatherDetailCard(
        title = "Wind",
        icon = Icons.Default.ArrowForward,
        modifier = modifier
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            weather.windSpeed?.let { speed ->
                DetailRow("Speed", "${speed.toInt()} km/h", isHighlight = true)
            }
            weather.windDirection?.let { direction ->
                DetailRow("Direction", "${direction}° ${getWindDirection(direction)}")
            }
        }
    }
}

@Composable
fun AtmosphericDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    WeatherDetailCard(
        title = "Atmospheric",
        icon = Icons.Default.Build,
        modifier = modifier
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            weather.pressure?.let { pressure ->
                DetailRow("Pressure", "${pressure.toInt()} hPa", isHighlight = true)
            }
            DetailRow("Humidity", "${weather.humidity}%")
            weather.visibility?.let { visibility ->
                DetailRow("Visibility", "${visibility.toInt()} km")
            }
            weather.cloudCover?.let { cloudCover ->
                DetailRow("Cloud cover", "${cloudCover}%")
            }
        }
    }
}

@Composable
fun PrecipitationDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    WeatherDetailCard(
        title = "Precipitation",
        icon = Icons.Default.Clear,
        modifier = modifier
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            weather.precipitationChance?.let { chance ->
                DetailRow("Chance", "${chance}%", isHighlight = true)
            }
            weather.precipitationAmount?.let { amount ->
                DetailRow("Amount", "${amount} mm")
            }
        }
    }
}

@Composable
fun UVIndexDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    WeatherDetailCard(
        title = "UV Index",
        icon = Icons.Default.Star,
        modifier = modifier
    ) {
        weather.uvIndex?.let { uvIndex ->
            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                DetailRow("UV Index", uvIndex.toString(), isHighlight = true)
                DetailRow("Category", getUVCategory(uvIndex))
                Text(
                    text = getUVAdvice(uvIndex),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.8f)
                )
            }
        }
    }
}

@Composable
fun AirQualityDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    weather.airQuality?.let { airQuality ->
        WeatherDetailCard(
            title = "Air Quality",
            icon = Icons.Default.ArrowForward,
            modifier = modifier
        ) {
            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                DetailRow("AQI", airQuality.aqi.toString(), isHighlight = true)
                DetailRow("Category", airQuality.category.name.replace("_", " "))
                airQuality.pm25?.let { pm25 ->
                    DetailRow("PM2.5", "${pm25} μg/m³")
                }
                airQuality.pm10?.let { pm10 ->
                    DetailRow("PM10", "${pm10} μg/m³")
                }
            }
        }
    }
}

@Composable
fun SunMoonDetailCard(
    weather: Weather,
    modifier: Modifier = Modifier
) {
    WeatherDetailCard(
        title = "Sun & Moon",
        icon = Icons.Default.Home,
        modifier = modifier
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            weather.sunrise?.let { sunrise ->
                DetailRow("Sunrise", sunrise, isHighlight = true)
            }
            weather.sunset?.let { sunset ->
                DetailRow("Sunset", sunset)
            }
            weather.moonPhase?.let { moonPhase ->
                DetailRow("Moon phase", moonPhase)
            }
        }
    }
}

@Composable
private fun DetailRow(
    label: String,
    value: String,
    isHighlight: Boolean = false,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.8f)
        )
        Text(
            text = value,
            style = if (isHighlight) MaterialTheme.typography.titleMedium else MaterialTheme.typography.bodyMedium,
            fontWeight = if (isHighlight) FontWeight.SemiBold else FontWeight.Normal,
            color = if (isHighlight) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

private fun getWindDirection(degrees: Int): String {
    return when (((degrees + 22.5) / 45).toInt()) {
        0, 8 -> "N"
        1 -> "NE"
        2 -> "E"
        3 -> "SE"
        4 -> "S"
        5 -> "SW"
        6 -> "W"
        7 -> "NW"
        else -> "N"
    }
}

private fun getUVCategory(uvIndex: Int): String {
    return when (uvIndex) {
        in 0..2 -> "Low"
        in 3..5 -> "Moderate"
        in 6..7 -> "High"
        in 8..10 -> "Very High"
        else -> "Extreme"
    }
}

private fun getUVAdvice(uvIndex: Int): String {
    return when (uvIndex) {
        in 0..2 -> "Minimal sun protection required"
        in 3..5 -> "Seek shade during midday hours"
        in 6..7 -> "Sun protection essential"
        in 8..10 -> "Extra protection needed"
        else -> "Take all precautions"
    }
}

File: iosApp/iosApp/Views/Molecules/WeatherDetailCard.swift
import SwiftUI
import shared

struct WeatherDetailCard: View {
    let title: String
    let icon: String
    let content: AnyView
    
    init<Content: View>(title: String, icon: String, @ViewBuilder content: () -> Content) {
        self.title = title
        self.icon = icon
        self.content = AnyView(content())
    }
    
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(spacing: 8) {
                Text(icon)
                    .font(.title2)
                Text(title)
                    .font(.headline)
                    .fontWeight(.semibold)
                    .foregroundColor(.primary)
                Spacer()
            }
            
            content
        }
        .padding(16)
        .background(
            RoundedRectangle(cornerRadius: 16)
                .fill(Color(.systemGray6))
        )
        .shadow(color: .black.opacity(0.05), radius: 4, x: 0, y: 2)
    }
}

struct TemperatureDetailCard: View {
    let weather: Weather
    
    var body: some View {
        WeatherDetailCard(title: "Temperature", icon: "🌡️") {
            VStack(spacing: 8) {
                if let current = weather.temperatureCurrent {
                    DetailRow(label: "Current", value: "\(Int(current))°C", isHighlight: true)
                }
                if let feelsLike = weather.feelsLike {
                    DetailRow(label: "Feels like", value: "\(Int(feelsLike))°C")
                }
                DetailRow(label: "High", value: "\(Int(weather.temperatureHigh))°C")
                DetailRow(label: "Low", value: "\(Int(weather.temperatureLow))°C")
                if let dewPoint = weather.dewPoint {
                    DetailRow(label: "Dew point", value: "\(Int(dewPoint))°C")
                }
            }
        }
    }
}

struct WindDetailCard: View {
    let weather: Weather
    
    var body: some View {
        WeatherDetailCard(title: "Wind", icon: "💨") {
            VStack(spacing: 8) {
                if let speed = weather.windSpeed {
                    DetailRow(label: "Speed", value: "\(Int(speed)) km/h", isHighlight: true)
                }
                if let direction = weather.windDirection {
                    DetailRow(label: "Direction", value: "\(direction)° \(getWindDirection(direction))")
                }
            }
        }
    }
    
    private func getWindDirection(_ degrees: Int32) -> String {
        let index = Int((Double(degrees) + 22.5) / 45) % 8
        let directions = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]
        return directions[index]
    }
}

struct AtmosphericDetailCard: View {
    let weather: Weather
    
    var body: some View {
        WeatherDetailCard(title: "Atmospheric", icon: "📊") {
            VStack(spacing: 8) {
                if let pressure = weather.pressure {
                    DetailRow(label: "Pressure", value: "\(Int(pressure)) hPa", isHighlight: true)
                }
                DetailRow(label: "Humidity", value: "\(weather.humidity)%")
                if let visibility = weather.visibility {
                    DetailRow(label: "Visibility", value: "\(Int(visibility)) km")
                }
                if let cloudCover = weather.cloudCover {
                    DetailRow(label: "Cloud cover", value: "\(cloudCover)%")
                }
            }
        }
    }
}

struct PrecipitationDetailCard: View {
    let weather: Weather
    
    var body: some View {
        WeatherDetailCard(title: "Precipitation", icon: "💧") {
            VStack(spacing: 8) {
                if let chance = weather.precipitationChance {
                    DetailRow(label: "Chance", value: "\(chance)%", isHighlight: true)
                }
                if let amount = weather.precipitationAmount {
                    DetailRow(label: "Amount", value: "\(amount, specifier: "%.1f") mm")
                }
            }
        }
    }
}

struct UVIndexDetailCard: View {
    let weather: Weather
    
    var body: some View {
        WeatherDetailCard(title: "UV Index", icon: "☀️") {
            if let uvIndex = weather.uvIndex {
                VStack(spacing: 8) {
                    DetailRow(label: "UV Index", value: "\(uvIndex)", isHighlight: true)
                    DetailRow(label: "Category", value: getUVCategory(Int(uvIndex)))
                    Text(getUVAdvice(Int(uvIndex)))
                        .font(.caption)
                        .foregroundColor(.secondary)
                        .multilineTextAlignment(.leading)
                }
            }
        }
    }
    
    private func getUVCategory(_ uvIndex: Int) -> String {
        switch uvIndex {
        case 0...2: return "Low"
        case 3...5: return "Moderate"
        case 6...7: return "High"
        case 8...10: return "Very High"
        default: return "Extreme"
        }
    }
    
    private func getUVAdvice(_ uvIndex: Int) -> String {
        switch uvIndex {
        case 0...2: return "Minimal sun protection required"
        case 3...5: return "Seek shade during midday hours"
        case 6...7: return "Sun protection essential"
        case 8...10: return "Extra protection needed"
        default: return "Take all precautions"
        }
    }
}

struct AirQualityDetailCard: View {
    let weather: Weather
    
    var body: some View {
        if let airQuality = weather.airQuality {
            WeatherDetailCard(title: "Air Quality", icon: "🌫️") {
                VStack(spacing: 8) {
                    DetailRow(label: "AQI", value: "\(airQuality.aqi)", isHighlight: true)
                    DetailRow(label: "Category", value: airQuality.category.name.replacingOccurrences(of: "_", with: " "))
                    if let pm25 = airQuality.pm25 {
                        DetailRow(label: "PM2.5", value: "\(pm25, specifier: "%.1f") μg/m³")
                    }
                    if let pm10 = airQuality.pm10 {
                        DetailRow(label: "PM10", value: "\(pm10, specifier: "%.1f") μg/m³")
                    }
                }
            }
        }
    }
}

struct SunMoonDetailCard: View {
    let weather: Weather
    
    var body: some View {
        WeatherDetailCard(title: "Sun & Moon", icon: "🌅") {
            VStack(spacing: 8) {
                if let sunrise = weather.sunrise {
                    DetailRow(label: "Sunrise", value: sunrise, isHighlight: true)
                }
                if let sunset = weather.sunset {
                    DetailRow(label: "Sunset", value: sunset)
                }
                if let moonPhase = weather.moonPhase {
                    DetailRow(label: "Moon phase", value: moonPhase)
                }
            }
        }
    }
}

struct DetailRow: View {
    let label: String
    let value: String
    let isHighlight: Bool
    
    init(label: String, value: String, isHighlight: Bool = false) {
        self.label = label
        self.value = value
        self.isHighlight = isHighlight
    }
    
    var body: some View {
        HStack {
            Text(label)
                .font(.body)
                .foregroundColor(.secondary)
            Spacer()
            Text(value)
                .font(isHighlight ? .headline : .body)
                .fontWeight(isHighlight ? .semibold : .regular)
                .foregroundColor(isHighlight ? .primary : .secondary)
        }
    }
}

#Preview {
    VStack(spacing: 16) {
        TemperatureDetailCard(weather: Weather(
            date: LocalDate(year: 2024, month: 1, dayOfMonth: 15),
            condition: WeatherCondition.clear,
            temperatureHigh: 25.0,
            temperatureLow: 15.0,
            temperatureCurrent: 22.0,
            humidity: 65,
            icon: "01d",
            description: "Clear sky",
            pressure: 1013.2,
            windSpeed: 12.5,
            windDirection: 225,
            visibility: 10.0,
            uvIndex: 6,
            precipitationChance: 20,
            precipitationAmount: 0.5,
            cloudCover: 15,
            feelsLike: 24.0,
            dewPoint: 12.0,
            sunrise: "06:30",
            sunset: "18:45",
            moonPhase: "Waxing Crescent",
            airQuality: nil
        ))
        
        HStack(spacing: 16) {
            WindDetailCard(weather: Weather(
                date: LocalDate(year: 2024, month: 1, dayOfMonth: 15),
                condition: WeatherCondition.clear,
                temperatureHigh: 25.0,
                temperatureLow: 15.0,
                humidity: 65,
                icon: "01d",
                description: "Clear",
                windSpeed: 12.5,
                windDirection: 225
            ))
            .frame(maxWidth: .infinity)
            
            AtmosphericDetailCard(weather: Weather(
                date: LocalDate(year: 2024, month: 1, dayOfMonth: 15),
                condition: WeatherCondition.clear,
                temperatureHigh: 25.0,
                temperatureLow: 15.0,
                humidity: 65,
                icon: "01d",
                description: "Clear",
                pressure: 1013.2,
                visibility: 10.0,
                cloudCover: 15
            ))
            .frame(maxWidth: .infinity)
        }
    }
    .padding()
}